### PR TITLE
Add dot plot and percentile lines to CDF plot on one-push page

### PR DESCRIPTION
**Pull request by carolksun**
This PR adds a dot plot and lines representing percentiles on top of the 
existing static CDF plot. The dot plot places the dots in a non-overlapping 
method that displays the quantity and raw duration of the completed pushes. 
The percentile lines are placed at the 10%, 50% and 90% values, if visually 
possible. If they overlap, then the 10% and 90% lines are moved outward until 
there is enough space. If there is no space, then only the 50% line appears on
the graph. Additionally, there is a line that appears on the graph of the 
completed pushes that represent how the current push compares to all the 
other pushes.

### CODE REVIEW

**Comment by carolksun** [comment]
Current version: https://pr31-73b0d74-dot-step189-2020.uc.r.appspot.com/
Example completed push 1.7k: https://pr31-73b0d74-dot-step189-2020.uc.r.appspot.com/c65c37c6e1fb/@20200805-130723.258753

**Comment by carolksun** [comment]
Current version: https://pr31-2272e2b-dot-step189-2020.uc.r.appspot.com/
Example completed push 1.7k: https://pr31-2272e2b-dot-step189-2020.uc.r.appspot.com/c65c37c6e1fb/@20200805-130723.258753

**Comment by carolksun** [comment]
Current version: https://pr31-8ed6ce5-dot-step189-2020.uc.r.appspot.com/
Example completed push 1.7k: https://pr31-8ed6ce5-dot-step189-2020.uc.r.appspot.com/c65c37c6e1fb/@20200805-130723.258753

File: web/src/app/components/cdf/cdf.component.ts [hunk 1]
```diff
@@ -143,25 +89,24 @@ export class CDFComponent implements AfterViewInit {
       .attr('width', elementWidth)
       .attr('height', elementHeight);
 
-    const cdfChart = this.svg
+    const CDFChart = this.svg
```
Comment (yinggao) [inline web/src/app/components/cdf/cdf.component.ts:92]: Why capitalizing CDFChart? We should only do `CDFChart` when the first letter `c` needs to be capitalized.

File: web/src/app/components/cdf/cdf.utils.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    return [percentileLines[1]].map(d => ({
+      duration: getXforPercentage(data, d),
+      probability: d} as Item));
+  }
+  let quantiles = percentileLines.map(d => ({
+    duration: getXforPercentage(data, d),
+    probability: d} as Item));
+
+  const differences = [xScale(quantiles[1].duration - quantiles[0].duration), xScale(quantiles[2].duration - quantiles[1].duration)];
+  if (differences[0] < 15 || differences[1] < 15) {
```
Comment (yinggao) [inline web/src/app/components/cdf/cdf.utils.ts:122]: Prefer to define a const for the magic number 15.

File: web/src/app/components/cdf/cdf.utils.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+ currentPushLine
+   .append('text')
+   .attr('text-anchor', 'middle')
+   .attr('x', xScale(duration))
+   .attr('y', 0)
+   .attr('id', 'current-line-text')
+   .attr('font-size', '12px')
+   .text('Current Push');
+   */
+}
```
Comment (yinggao) [inline web/src/app/components/cdf/cdf.utils.ts:265]: Unused code?

Comment (carolksun) [reply web/src/app/components/cdf/cdf.utils.ts:265]: I was saving the code to create the full vertical line but now that we've settle on the arrow version, I can delete it!

**Comment by carolksun** [comment]
Current version: https://pr31-d11ecf1-dot-step189-2020.uc.r.appspot.com/
Example completed push 1.7k: https://pr31-d11ecf1-dot-step189-2020.uc.r.appspot.com/c65c37c6e1fb/@20200805-130723.258753

**Comment by carolksun** [comment]
Current version: https://pr31-94025a5-dot-step189-2020.uc.r.appspot.com/
Example completed push 1.7k: https://pr31-94025a5-dot-step189-2020.uc.r.appspot.com/c65c37c6e1fb/@20200805-130723.258753
Example completed push 83: https://pr31-94025a5-dot-step189-2020.uc.r.appspot.com/7f4535707267/@20190610-170000.029625

**Comment by carolksun** [comment]
Current version: https://pr31-e8cee86-dot-step189-2020.uc.r.appspot.com/
Example completed push 1.7k: https://pr31-e8cee86-dot-step189-2020.uc.r.appspot.com/c65c37c6e1fb/@20200805-130723.258753
Example completed push 83: https://pr31-e8cee86-dot-step189-2020.uc.r.appspot.com/7f4535707267/@20190610-170000.029625

File: web/src/app/components/cdf/cdf.utils.ts [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+ * increasing duration, the resulting probability is the probability of the
+ * Item with a duration directly less than the given duration parameter.
+ *
+ * @param data Array of Items sorted by increasing duration
+ * @param duration A duration value between the min and max of all durations
+ * @return A probability value between 0 and 1 (exclusive)
+ */
+export function getProbabilityForDuration(
+                  data: Item[],
+                  duration: number): number {
```
Comment (razvanm) [inline web/src/app/components/cdf/cdf.utils.ts:90]: These indentation doesn't seem to follow the guidelines from [1]. Same in the other similar places.

[1] https://google.github.io/styleguide/jsguide.html#formatting-function-arguments

File: web/src/app/components/cdf/cdf.utils.ts [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+ * order, if placing a dot overlaps with previously placed dots, then increase
+ * the y position to dodge that dot. This loops until the (x, y) position of the
+ * does not overlap with any other dots.
+ *
+ * @param radius radius of dots
+ * @param xScale d3 function that applies a scaling factor on raw x values to
+ * correctly place them on the graph
+ * @param xVals array of durations in sorted order
+ * @return Array of numbers representing the y position of the dot at duration
+ * value maintaining the same indices as the xVals
```
Comment (razvanm) [inline web/src/app/components/cdf/cdf.utils.ts:168]:
```suggestion
 * @param radius radius of dots
 * @param xScale d3 function that applies a scaling factor on raw x values to
 *     correctly place them on the graph
 * @param xVals array of durations in sorted order
 * @return Array of numbers representing the y position of the dot at duration
 *     value maintaining the same indices as the xVals
```

The continuation should use and indentation of 4 spaces. Same comment for the other similar places.

Reference: https://google.github.io/styleguide/jsguide.html#jsdoc-line-wrapping

**Pull request merged**